src/poprox_recommender/components/diversifiers/locality_calibration.py

```diff
@@ -25,13 +47,234 @@ def __call__(self, candidate_articles: ArticleSet, interest_profile: InterestPro
         article_indices = self.calibration(
             article_scores,
             candidate_articles.articles,
+            normalized_topic_prefs,
             normalized_locality_prefs,
-            self.theta,
+            self.theta_topic,
+            self.theta_local,
             topk=self.num_slots,
         )
-        return ArticleSet(articles=[candidate_articles.articles[int(idx)] for idx in article_indices])
 
-    def add_article_to_categories(self, rec_categories, article):
-        locality_list = extract_locality(article)
-        for locality in locality_list:
-            rec_categories[locality] = rec_categories.get(locality, 0) + 1
+        recommended_list = text_generation(candidate_articles.articles, article_indices, interest_profile)
+        return ArticleSet(articles=recommended_list)
```

Comment:
This seems like a whole separate component to me, since the text generation gets applied to every selected article and isn't otherwise coupled to the calibration code. I'd probably use the `article_indices` and `candidate_articles` to build and return an `ArticleSet` here: 
```suggestion
        return ArticleSet(articles=[candidate_articles.articles[idx] for idx in article_indices])
```
and then wire that output into a new component (called something like `ContextGenerator`?) that does the text generation part. More on this in the `text_generation` code below.

Reply:
This was a challenge I was referring to a few weeks ago over Slack. Text generation shouldn't be applied to every article, so we'd need to set a flag of sorts if we want to pull out this functionality to another component.

Reply:
Yeah, you can set whatever properties you want on `ArticleSet`, so you could add a list or tensor with boolean flags to indicate whether you want it applied or not and that'll get sent along to subsequent components. I might be missing something but I don't see where in the current code the equivalent selection is happening—it appears to be applying LLM generated text to all selected/recommended articles?

Reply:
Have moved `text_generation` out of the` __call__` method.